Enable generate runtime config and deps.files

wli3 opened the pull request:
Please review it **commit by commit**.

Fix https://github.com/dotnet/sdk/issues/3889

Commit:
Add test asset

Commit:
Block self-contained with C++/CLI project

Commit:
Add enable Deps json file without asset file

Commit:
Enable Generate runtime config without asset file

Commit:
Fix no Runtime config generated

When asset file is missing. By adding correct incremental build with property hased

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -222,6 +224,34 @@ Copyright (c) .NET Foundation. All rights reserved.
 
+  <!-- To achieve incremental build with property change. When any property changes, WriteOnlyWhenDifferent will be triggered to write cache file.
+  And the cache file's timestamp will be later, and it then triggers the incremental build.-->
+  <Target Name="_GenerateRuntimeConfigurationFilesInputCache" DependsOnTargets="ResolveAssemblyReferences">
```
need discussion. maybe over kill, but that should be the "right thing"

nguerrera replied:
Seems OK. 

But why does this depend on ResolveAssemblyReferences? I don't see anything from what it does being used here.

runtimeconfig is so small, that I sort of wonder if it wound't be simpler, easier to maintain, or more efficient to just generate it and write the whole file only if different rather than any hashing that might need to be kept up to date with new properties.

wli3 replied:
Good idea. The hash is probably bigger than the file. I reverted that commit and added https://github.com/dotnet/sdk/pull/4063/commits/31cec9dcea3ada9732bda2e46f64f6e997d17c90 instead.

Commit:
Fix bad merge

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs:
```diff
@@ -138,8 +154,13 @@ private void WriteDepsFile(string depsFilePath)
 
+            // If there is a generated asset file. The projectContext will have project reference.
+            // So remove it from directReferences to avoid duplication
+            var projectReferenceExistedInProjectContext = projectContext != null;
```
>projectReferenceExistedInProjectContext [](start = 16, length = 39)

Naming suggestion: `projectContextHasProjectReferences`

Elsewhere it was confusing me that this seemed to be referring to a single project refer rather than the general set of all project references.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenADependencyContextBuilder.cs:
```diff
@@ -52,6 +52,7 @@ public class GivenADependencyContextBuilder
                 ReferenceInfo.CreateDirectReferenceInfos(
                     referencePaths ?? new ITaskItem[] { },
                     referenceSatellitePaths ?? new ITaskItem[] { },
+                    projectReferenceExistedInProjectContext: false,
```
Naming suggestion: `projectContextHasProjectReferences` (here and wherever else this name is used.

Elsewhere it was confusing me that this seemed to be referring to a single project reference rather than the general set of all project references.

wli3 replied:
renamed

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAGenerateRuntimeConfigurationFiles.cs:
```diff
@@ -0,0 +1,248 @@
+    }
+  }
+}");
+            File.Exists(_runtimeConfigDevPath).Should().BeFalse("No nuget involved, so no extra probing path");
```
Sounds fine to me.

nguerrera commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAGenerateRuntimeConfigurationFiles.cs:
```diff
@@ -0,0 +1,248 @@
+            }
+        }
+
+        private class MockBuildEngine4 : MockBuildEngine, IBuildEngine4
```
Do we not have one of these somewhere that can be reused?

wli3 replied:
That's the only place we use MockBuildEngine**4** since we have more logic about the cache has dependency on it

wli3 commented:
@nguerrera https://github.com/dotnet/sdk/pull/4063/commits/31cec9dcea3ada9732bda2e46f64f6e997d17c90 is the new commit for fix missing runtimeconfig. I resolved your comments so far.

wli3 commented:
@nguerrera 31cec9d is the new commit for fix missing runtimeconfig. I resolved your comments so far.

nguerrera commented:
I decided to check on the perf here. It seems GenerateRuntimeConfigurationFiles is  a noticeable part of incremental build when assets file has changed.

You can compare `dotnet build /clp:PerformanceSummary` vs. `touch obj\projects.assets.json && donet build /clp:PerformanceSummary`

I'm doing this on a vanilla `dotnet new console` and seeing GenerateRutnimeConfigurationFiles move to near the most expensive.

```
      38 ms  GenerateDepsFile                           1 calls
       74 ms  GenerateRuntimeConfigurationFiles          1 calls
      100 ms  MsBuild                                    7 calls
      102 ms  RestoreTask                                1 calls
```

I am surprised by this result, but it might invalidate my idea of doing it this way.
```

nguerrera commented:
What do you think about treating the perfect up-to-date-ness of runtime config as a separate change from this one?

wli3 commented:
@nguerrera good idea

Commit:
Address code review

- Rename to projectContextHasProjectReferences

- Move DependsOnTargets=_DefaultMicrosoftNETPlatformLibrary

- Revert unnecessary dependency

wli3 commented:
Use back incremental build fix instead. Log this for 5.0 https://github.com/dotnet/sdk/issues/4136

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -570,7 +571,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     ============================================================
     -->
   <Target Name="_ComputeNETCoreBuildOutputFiles"
-          DependsOnTargets="_GetAppHostPaths;_GetComHostPaths;_GetIjwHostPaths;GenerateBuildRuntimeConfigurationFiles"
```
Reverted this

wli3 commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:
```diff
@@ -261,7 +262,7 @@ Copyright (c) .NET Foundation. All rights reserved.
     -->
 
   <Target Name="GenerateBuildRuntimeConfigurationFiles"
-          DependsOnTargets="_DefaultMicrosoftNETPlatformLibrary;_GenerateRuntimeConfigurationFilesInputCache"
```
Move dependency on _DefaultMicrosoftNETPlatformLibrary to _GenerateRuntimeConfigurationFilesInputCache. So _GenerateRuntimeConfigurationFilesInputCache can get the same property as before. Some property need to be set by a target before _GenerateRuntimeConfigurationFilesInputCache runs
